joint_trajectory_controller/src/joint_trajectory_controller.cpp

```diff
@@ -374,13 +530,117 @@ void JointTrajectoryController::publish_state()
     last_state_publish_time_ = lifecycle_node_->now();
 
     state_publisher_->msg_.header.stamp = last_state_publish_time_;
-    // TODO(ddengster): Fill in the rest of the state data.
-    // Port it when we're ready to put in the other trajectory code (ros2_controllers PR #26).
+    state_publisher_->msg_.desired.positions = desired_state.positions;
+    state_publisher_->msg_.desired.velocities = desired_state.velocities;
+    state_publisher_->msg_.desired.accelerations = desired_state.accelerations;
+    state_publisher_->msg_.actual.positions = current_state.positions;
+    state_publisher_->msg_.actual.velocities = current_state.velocities;
+    state_publisher_->msg_.error.positions = state_error.positions;
+    state_publisher_->msg_.error.velocities = state_error.velocities;
 
     state_publisher_->unlockAndPublish();
   }
 }
 
+rclcpp_action::GoalResponse JointTrajectoryController::goal_callback(
+  const rclcpp_action::GoalUUID & uuid,
+  std::shared_ptr<const FollowJTrajAction::Goal> goal)
+{
+  RCLCPP_INFO(lifecycle_node_->get_logger(), "Received new action goal");
+  (void)uuid;
+
+  // Precondition: Running controller
+  if (is_halted) {
+    RCLCPP_ERROR(
+      lifecycle_node_->get_logger(),
+      "Can't accept new action goals. Controller is not running.");
+    return rclcpp_action::GoalResponse::REJECT;
+  }
+
+  // If partial joints goals are not allowed, goal should specify all controller joints
+  if (!allow_partial_joints_goal_) {
+    if (goal->trajectory.joint_names.size() != joint_names_.size()) {
+      RCLCPP_ERROR(
+        lifecycle_node_->get_logger(),
+        "Joints on incoming goal don't match the controller joints.");
+      return rclcpp_action::GoalResponse::REJECT;
+    }
+  }
+
+  RCLCPP_INFO(lifecycle_node_->get_logger(), "Accepted new action goal");
+  return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
+}
+
+rclcpp_action::CancelResponse JointTrajectoryController::cancel_callback(
+  const std::shared_ptr<rclcpp_action::ServerGoalHandle<FollowJTrajAction>> goal_handle)
+{
+  RCLCPP_INFO(lifecycle_node_->get_logger(), "Got request to cancel goal");
+  (void)goal_handle;
+
+  // Check that cancel request refers to currently active goal (if any)
+  if (rt_active_goal_ && rt_active_goal_->gh_ == goal_handle) {
+    // Controller uptime
+    // Enter hold current position mode
+    set_hold_position();
```

Comment:
does this have to be locked in any way?

Reply:
there's a lock_guard in set_hold_position